feat: Members service

kimseungbin opened the pull request:
# Description

- 멤버 서비스의 주요 기능 추가

# Todos

- [x] Test case 작성
- [x] findAll
- [x] findOne
- [x] create
- [x] update
- [x] delete

- Close #11 

# 질문들

1. Update 시 unique 한 값을 가지는 column 의 대한 검사를 transaction 을 포함하여 find 한 후에 update 하는 것 vs. update 시에 한 번에 처리하는 것
2. Mocking vs stubbing
3. https://docs.nestjs.com/recipes/automock
4. controller 나 repository 단위의 테스트 필요 여부. 만약 controller 를 테스트하게된다면 테스트 대상은 무엇인지.

kimseungbin committed:
test: Add members service test

kimseungbin committed:
test: Add cases for findAll in members service

kimseungbin committed:
test: Add cases for findOne in members service

kimseungbin committed:
test: Add cases for create in members service

kimseungbin committed:
test: Add cases for findAll in members service

kimseungbin committed:
test: Add cases for delete in members service

kimseungbin commented:
> Update 시 unique 한 값을 가지는 column 의 대한 검사를 transaction 을 포함하여 find 한 후에 update 하는 것 vs. update 시에 한 번에 처리하는 것

App Db 리소스 중에 덜 비싼 것으로 처리
https://stackoverflow.com/questions/67197276/how-do-i-make-combination-of-3-columns-in-typeorm-postgres-unique

DB 를 scale out 하는 것이 더 어려우므로

kimseungbin committed:
test: Add `should return the array of members` case

kimseungbin committed:
feat: Implement min of findAll members to run test

kimseungbin committed:
feat: Implement `should return the array of members` test

kimseungbin committed:
test: Fix expected array resulting in 0 length

kimseungbin committed:
test: Add `should throw HttpException No Content`

kimseungbin committed:
test: Add `when querying without any column` case

kimseungbin committed:
test: Add `when querying by a value of the specific column` case

kimseungbin committed:
feat: Implement min of findOne member to run test

kimseungbin committed:
test: Remove 'when querying by a value of the specific column' case

As finding only one member by other than ID would be unnecessary, it will split into two test cases. The first is findOneById and the second is find phone number with autocomplete feature.

kimseungbin committed:
test: Add findAllByPhone test cases

kimseungbin committed:
test: Add pagination case

kimseungbin committed:
test: Add findOneBy test

kimseungbin committed:
fix: Change type of member id number -> string

UUID is string not number

kimseungbin committed:
feat: member service findOneBy

kimseungbin committed:
test: Remove unused findOne member test

kimseungbin committed:
test: Add findAll members test cases for 'when handling deleted members'

kimseungbin committed:
test: Add 'should return only non-deleted members by default' case

kimseungbin committed:
test: Add 'should return all members, including deleted ones, if parameter `status` is "all"' case

kimseungbin committed:
feat: Add status check in findAll members service

By taking status as a parameter, it is possible to look for active, deleted and all users.

kimseungbin committed:
feat: Refactor member status test

Previously, the member status test didn't effectively test the application logic; it only verified the functionality of the Jest mock. With the introduction of method spying, the test now accurately validates the effectiveness of the if-statement logic.

kimseungbin committed:
docs: Add Test-Driven Development section to README

Expanded the README with a comprehensive new section on Test-Driven Development (TDD). This includes detailed information about the role of TDD in the project, as well as examples and explanations of test cases that utilize mocking and spy techniques.

kimseungbin committed:
feat: Add CreateMemberDto with validation rules

This commit adds a new Data Transfer Object (DTO) for creating a new member. It also includes validation rules using class-validator for the input fields such as the member's first name, last name, nickname, and phone number. This will ensure that all input data meets the specified criteria before creating a new member.

kimseungbin committed:
test: Implement create member test in MemberService

kimseungbin committed:
feat: Implement 'create' functionality in MembersService

kimseungbin committed:
test: Remove unused create member test case.

Since request body validation should not be handled in service layer.

kimseungbin committed:
feat: Add exceptions library

kimseungbin committed:
feat: Add DuplicateMemberException class

kimseungbin committed:
feat: Add DuplicateMemberException handling to member creation test

kimseungbin committed:
feat: Add database configuration

kimseungbin committed:
feat: Add error handling to member creation

eunkyui commented on apps/members/src/members.service.ts:
```diff
@@ -1,16 +1,63 @@
+		return members
+	}
+
+	async findOneBy(column: keyof Member, value: any): Promise<Member> {
```
findOneBy 를 쓰기 부적절한 column 도 많을 것 같네요

eunkyui commented on apps/members/src/members.service.ts:
```diff
@@ -1,16 +1,63 @@
+		try {
+			return await this.memberRepository.save(member)
+		} catch (error) {
+			if (this.configService.get('database') === 'postgres') {
```
DB insert/update 시 unique 필드 중복으로 error 날 때 체크하는건 member service 외에도 많이 쓰일 것 같은데 이런 기능은 별도로 빼봐도 좋을거 같네요

kimseungbin replied:
별개의 함수로 일단 분리하였습니다. 563bfb22

kimseungbin committed:
test: Merge duplicate member tests

As validating duplicate logic is handled within the database, service layer cannot distinguish the cause of the duplicate

kimseungbin committed:
feat: Add UpdateMemberDto class

kimseungbin committed:
feat: Add MemberDto class

A new abstract class member.dto.ts has been added to the members module. This class contains optional string attributes like firstName, lastName, nickname and phone. Validations have also been implemented for these attributes.

kimseungbin committed:
feat: Refactor create and update member DTOs to use member DTO

kimseungbin committed:
test: Add test cases for update member role

kimseungbin committed:
test: Add update member test

kimseungbin committed:
feat: Add update method in MembersService

kimseungbin committed:
test: Add member service duplicate update case

kimseungbin committed:
feat: Add error handling to member update service

kimseungbin committed:
refactor: Error handling in members service

kimseungbin committed:
test: Remove unused member update tests

kimseungbin committed:
test: Add update member role case

kimseungbin committed:
feat: Implement members service updateRole method

kimseungbin committed:
test: Add MemberNotFoundException test case

kimseungbin committed:
test: Add NonSequentialRoleUpdateException and test

kimseungbin committed:
feat: Refactor MemberRole to Role

kimseungbin committed:
test: Implement updateRole test

kimseungbin committed:
feat: Implement members service update role

kimseungbin committed:
test: Add members service delete test case

kimseungbin committed:
feat: Add members service delete method

kimseungbin committed:
feat: Add MemberRedundantDeletionException case

kimseungbin committed:
feat: Update members service delete to set role to null

kimseungbin committed:
test: Mock new Date to match test and service logic

kimseungbin committed:
test: Add config service dependency in members controller test
